src/vec.rs

```diff
@@ -594,7 +591,7 @@ impl<T> EcoVec<T> {
     fn data(&self) -> *const T {
         // If sentinel pointer isn't a valid pointer of T, fabricate one.
         // This only happens if `len == 0`, so a dangling pointer is fine.
-        if mem::align_of::<T>() > mem::align_of::<Sentinel>() && !self.is_allocated() {
+        if mem::align_of::<T>() > mem::align_of::<u8>() && !self.is_allocated() {
```

Comment:
Also, this does slow down `as_slice` in practice because the align of a typical `T` is often larger than that of `u8`, leading to this extra branch. Whereas sentinel has intentionally large alignment, so that this case will be optimized away for `mem::align_of::<T>() <= 32`.

Reply:
Your reasoning for the header could give us a bit more leeway here though. Null can't be a valid allocation, so a dangling pointer with the value of the header size would work just as well as a dangling pointer with value 1 (if we can get it to work safely at all). On 64-bit, this branch would be eliminated for types with 16 byte alignment, which would be okay. And 8 byte on 32-bit would be okay too I guess.